crates/bevy_core/src/time/timer.rs

```diff
@@ -4,39 +4,58 @@ use bevy_property::Properties;
 use std::time::Duration;
 
 /// Tracks elapsed time. Enters the finished state once `duration` is reached.
+///
+/// Non repeating timers will stop tracking and stay in the finished state until reset.
+/// Repeating timers will only be in the finished state on each tick `duration` is reached or exceeded, and can still be reset at any given point.
 #[derive(Clone, Debug, Default, Properties)]
 pub struct Timer {
     pub elapsed: f32,
     pub duration: f32,
     pub finished: bool,
+    /// Stores the last tick's finished state. Is checked against in just_finished.
+    prev_finished: bool,
+    pub repeating: bool,
 }
 
 impl Timer {
-    pub fn new(duration: Duration) -> Self {
+    pub fn new(duration: Duration, repeating: bool) -> Self {
         Timer {
             duration: duration.as_secs_f32(),
+            repeating,
             ..Default::default()
         }
     }
 
-    pub fn from_seconds(seconds: f32) -> Self {
+    pub fn from_seconds(seconds: f32, repeating: bool) -> Self {
         Timer {
             duration: seconds,
+            repeating,
             ..Default::default()
         }
     }
 
     pub fn tick(&mut self, delta: f32) {
-        self.elapsed = (self.elapsed + delta).min(self.duration);
-        if self.elapsed >= self.duration {
-            self.finished = true;
+        if self.elapsed < self.duration {
```

Comment:
Wouldn't it cause just_finished to stay true even after ticks of it finishing? (as a non repeating timer)

Reply:
Yup, this way it defeats the purpose of `just_finished()` for non-repeating timers, but I think it could work:

```rust
self.just_finished = false;
self.finished = self.elapsed >= self.duration;
if self.finished && !self.just_finished {
  self.just_finished = true;
  if self.repeating {
    self.elapsed = self.elapsed % self.duration;
  }
}
```

Might need a few mre changes. I'll try around a bit. Having `just_finished` as a pub field does make it nicer/simpler than my current implementation.

Reply:
I think the following would work to remove `just_finished()` and replace `prev_finished` with `just_finished`:

```rust
    //...
    pub fn tick(&mut self, delta: f32) {
        let prev_finished = self.elapsed >= self.duration;
        if self.elapsed < self.duration {
            self.elapsed += delta;
        }

        self.finished = self.elapsed >= self.duration;
        self.just_finished = !prev_finished && self.finished;

        if self.repeating && self.finished {
            self.elapsed = self.elapsed % self.duration;
        }
    }

    pub fn reset(&mut self) {
        self.finished = false;
        self.just_finished = false;
        self.elapsed = 0.0;
    }
```

Will test in a bit.

Reply:
Looks good, the only thing I would change is the first if statement to:
`if !prev_finished {`